extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java

```diff
@@ -108,91 +115,79 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
-                parseSpec.getTimestampSpec().extractTimestamp(intGauge),
-                dimensions,
-                intGauge
-            ));
+            derivedMetricsList.add(intGauge);
             break;
           case SUMMARY_VALUE:
             // count
-            HashMap<String, Object> summaryCount = new HashMap<>();
+            Map<String, Object> summaryCount = new HashMap<>();
             summaryCount.putAll(labels);
             summaryCount.put(NAME, metric.getMetricDescriptor().getName() + SEPARATOR + "count");
             summaryCount.put(VALUE, point.getSummaryValue().getCount().getValue());
-            rows.add(new MapBasedInputRow(
-                parseSpec.getTimestampSpec().extractTimestamp(summaryCount),
-                dimensions,
-                summaryCount
-            ));
+            derivedMetricsList.add(summaryCount);
 
             // sum
-            HashMap<String, Object> summarySum = new HashMap<>();
+            Map<String, Object> summarySum = new HashMap<>();
             summarySum.putAll(labels);
             summarySum.put(NAME, metric.getMetricDescriptor().getName() + SEPARATOR + "sum");
             summarySum.put(VALUE, point.getSummaryValue().getSnapshot().getSum().getValue());
-            rows.add(new MapBasedInputRow(
-                parseSpec.getTimestampSpec().extractTimestamp(summarySum),
-                dimensions,
-                summarySum
-            ));
+            derivedMetricsList.add(summarySum);
 
             // TODO : Do we put percentiles into druid ?
             break;
           case DISTRIBUTION_VALUE:
             // count
-            HashMap<String, Object> distCount = new HashMap<>();
+            Map<String, Object> distCount = new HashMap<>();
             distCount.put(NAME, metric.getMetricDescriptor().getName() + SEPARATOR + "count");
             distCount.put(VALUE, point.getDistributionValue().getCount());
-            rows.add(new MapBasedInputRow(
-                parseSpec.getTimestampSpec().extractTimestamp(distCount),
-                dimensions,
-                distCount
-            ));
+            derivedMetricsList.add(distCount);
 
             // sum
-            HashMap<String, Object> distSum = new HashMap<>();
+            Map<String, Object> distSum = new HashMap<>();
             distSum.put(NAME, metric.getMetricDescriptor().getName() + SEPARATOR + "sum");
             distSum.put(VALUE, point.getDistributionValue().getSum());
-            rows.add(new MapBasedInputRow(
-                parseSpec.getTimestampSpec().extractTimestamp(distSum),
-                dimensions,
-                distSum
-            ));
+            derivedMetricsList.add(distSum);
             // TODO: How to handle buckets ?
             break;
         }
+
+        // Add druid rows based on derivedMetrics list.
+        for (Map<String, Object> derivedMetrics : derivedMetricsList) {
+          rows.add(new MapBasedInputRow(
```

Comment:
the derivedMetricsList seems a bit extraneous here. You seem to be using the list as a mechanism to avoid code duplication in each switch case, in which case extracting this into a function might make more sense. Given that this code is called extremely frequently, it would also avoid an unnecessary arraylist allocation.

tl'dr instead of `derivedMetrics.add()` call a function that does the `rows.add(...)` directly

Reply:
@xvrl Agree, shall change. WIll raise the PR, I have noted it.